### PR TITLE
HBASE-25003 Backport HBASE-24350 and HBASE-24779 to branch-2.2

### PR DESCRIPTION
**Pull request by tamasadami**

### CODE REVIEW

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   3m 49s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 2 new or modified test files.  |
||| _ branch-2.2 Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 15s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   3m 31s |  branch-2.2 passed  |
| +1 :green_heart: |  compile  |   1m 30s |  branch-2.2 passed  |
| +1 :green_heart: |  checkstyle  |   1m 53s |  branch-2.2 passed  |
| +1 :green_heart: |  shadedjars  |   4m 26s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   1m  5s |  branch-2.2 passed  |
| +0 :ok: |  spotbugs  |   3m 38s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   4m 33s |  branch-2.2 passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 13s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   3m 10s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 32s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 32s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   1m 52s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedjars  |   4m 23s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  hadoopcheck  |  27m 36s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 2.10.0 or 3.1.2 3.2.1.  |
| -1 :x: |  javadoc  |   0m 34s |  hbase-server generated 1 new + 1 unchanged - 0 fixed = 2 total (was 1)  |
| +1 :green_heart: |  findbugs  |   4m 54s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   0m 33s |  hbase-hadoop-compat in the patch passed.  |
| +1 :green_heart: |  unit  |   0m 38s |  hbase-hadoop2-compat in the patch passed.  |
| +1 :green_heart: |  unit  | 253m 14s |  hbase-server in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m 18s |  The patch does not generate ASF License warnings.  |
|  |   | 328m 36s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.12 Server=19.03.12 base: https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2372/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/2372 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 3744a00a006b 4.15.0-112-generic #113-Ubuntu SMP Thu Jul 9 23:41:39 UTC 2020 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-home/workspace/Base-PreCommit-GitHub-PR_PR-2372/out/precommit/personality/provided.sh |
| git revision | branch-2.2 / 88c9a38617 |
| Default Java | 1.8.0_181 |
| javadoc | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2372/1/artifact/out/diff-javadoc-javadoc-hbase-server.txt |
|  Test Results | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2372/1/testReport/ |
| Max. process+thread count | 2937 (vs. ulimit of 12500) |
| modules | C: hbase-hadoop-compat hbase-hadoop2-compat hbase-server U: . |
| Console output | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2372/1/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   2m 46s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 2 new or modified test files.  |
||| _ branch-2.2 Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 16s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   3m 34s |  branch-2.2 passed  |
| +1 :green_heart: |  compile  |   1m 34s |  branch-2.2 passed  |
| +1 :green_heart: |  checkstyle  |   1m 58s |  branch-2.2 passed  |
| +1 :green_heart: |  shadedjars  |   4m 45s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   1m 13s |  branch-2.2 passed  |
| +0 :ok: |  spotbugs  |   4m 16s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   5m 19s |  branch-2.2 passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 15s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   3m 36s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 50s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 50s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   2m  2s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedjars  |   4m 54s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  hadoopcheck  |  25m 31s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 2.10.0 or 3.1.2 3.2.1.  |
| +1 :green_heart: |  javadoc  |   1m  8s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   4m 35s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   0m 36s |  hbase-hadoop-compat in the patch passed.  |
| +1 :green_heart: |  unit  |   0m 42s |  hbase-hadoop2-compat in the patch passed.  |
| +1 :green_heart: |  unit  | 159m 36s |  hbase-server in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m 36s |  The patch does not generate ASF License warnings.  |
|  |   | 234m 25s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.12 Server=19.03.12 base: https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2372/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/2372 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux f19758d3d21b 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-home/workspace/Base-PreCommit-GitHub-PR_PR-2372/out/precommit/personality/provided.sh |
| git revision | branch-2.2 / 88c9a38617 |
| Default Java | 1.8.0_181 |
|  Test Results | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2372/2/testReport/ |
| Max. process+thread count | 3542 (vs. ulimit of 12500) |
| modules | C: hbase-hadoop-compat hbase-hadoop2-compat hbase-server U: . |
| Console output | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2372/2/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Review by jojochuang — APPROVED** [review]
I looked through all the changes. LGTM.

**Comment by jojochuang** [comment]
I'm sorry I'll have to revert this. looking at #2374 @ndimiduk has reservation around API change.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/MetricsSource.java [hunk 1]
```diff
@@ -71,8 +74,8 @@ public MetricsSource(String id) {
    * @param globalSourceSource Class to monitor global-scoped metrics
    */
   public MetricsSource(String id, MetricsReplicationSourceSource singleSourceSource,
-                       MetricsReplicationSourceSource globalSourceSource,
-                       Map<String, MetricsReplicationSourceSource> singleSourceSourceByTable) {
+                       MetricsReplicationGlobalSourceSource globalSourceSource,
```
Comment (jojochuang) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/MetricsSource.java:77]: Would it be possible to keep the existing API?

**Pull request merged**